机器学习

In [1]:
import pandas as pd

df = pd.read_csv('Musical_instruments_reviews.csv')
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10261 entries, 0 to 10260
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   reviewerID      10261 non-null  object
 1   asin            10261 non-null  object
 2   reviewerName    10234 non-null  object
 3   helpful         10261 non-null  object
 4   reviewText      10254 non-null  object
 5   overall         10261 non-null  int64 
 6   summary         10261 non-null  object
 7   unixReviewTime  10261 non-null  int64 
 8   reviewTime      10261 non-null  object
dtypes: int64(2), object(7)
memory usage: 721.6+ KB
None


In [2]:
df.head()

reviewerID        asin  \
0  A2IBPI20UZIR0U  1384719342   
1  A14VAT5EAX3D9S  1384719342   
2  A195EZSQDW3E21  1384719342   
3  A2C00NNG1ZQQG2  1384719342   
4   A94QU4C90B1AX  1384719342   

                                       reviewerName   helpful  \
0  cassandra tu "Yeah, well, that's just like, u...    [0, 0]   
1                                              Jake  [13, 14]   
2                     Rick Bennette "Rick Bennette"    [1, 1]   
3                         RustyBill "Sunday Rocker"    [0, 0]   
4                                     SEAN MASLANKA    [0, 0]   

                                          reviewText  overall  \
0  Not much to write about here, but it does exac...        5   
1  The product does exactly as it should and is q...        5   
2  The primary job of this device is to block the...        5   
3  Nice windscreen protects my MXL mic and preven...        5   
4  This pop filter is great. It looks and perform...        5   

                                 summary  unixReviewTime   reviewTime  
0                                   good      1393545600  02 28, 2014  
1                                   Jake      1363392000  03 16, 2013  
2                   It Does The Job Well      1377648000  08 28, 2013  
3          GOOD WINDSCREEN FOR THE MONEY      1392336000  02 14, 2014  
4  No more pops when I record my vocals.      1392940800  02 21, 2014

In [3]:
df.isnull().sum()

reviewerID         0
asin               0
reviewerName      27
helpful            0
reviewText         7
overall            0
summary            0
unixReviewTime     0
reviewTime         0
dtype: int64

In [4]:
df['reviewText'] = df['reviewText'].fillna("Missing")

In [6]:
df['reviews'] = df['summary'] + '. ' + df['reviewText']
df = df.drop(['reviewText', 'summary'], axis=1)
df.head()

reviewerID        asin  \
0  A2IBPI20UZIR0U  1384719342   
1  A14VAT5EAX3D9S  1384719342   
2  A195EZSQDW3E21  1384719342   
3  A2C00NNG1ZQQG2  1384719342   
4   A94QU4C90B1AX  1384719342   

                                       reviewerName   helpful  overall  \
0  cassandra tu "Yeah, well, that's just like, u...    [0, 0]        5   
1                                              Jake  [13, 14]        5   
2                     Rick Bennette "Rick Bennette"    [1, 1]        5   
3                         RustyBill "Sunday Rocker"    [0, 0]        5   
4                                     SEAN MASLANKA    [0, 0]        5   

   unixReviewTime   reviewTime  \
0      1393545600  02 28, 2014   
1      1363392000  03 16, 2013   
2      1377648000  08 28, 2013   
3      1392336000  02 14, 2014   
4      1392940800  02 21, 2014   

                                             reviews  
0  good. Not much to write about here, but it doe...  
1  Jake. The product does exactly as it should an...  
2  It Does The Job Well. The primary job of this ...  
3  GOOD WINDSCREEN FOR THE MONEY. Nice windscreen...  
4  No more pops when I record my vocals.. This po...

In [7]:
df['overall'].value_counts()

overall
5    6938
4    2084
3     772
2     250
1     217
Name: count, dtype: int64

In [11]:
def func(row):
    if row['overall'] >= 4.0:
        return 'positive'
    else:
        return 'negative'

df['sentiment'] = df.apply(func, axis=1)
df.head()

reviewerID        asin  \
0  A2IBPI20UZIR0U  1384719342   
1  A14VAT5EAX3D9S  1384719342   
2  A195EZSQDW3E21  1384719342   
3  A2C00NNG1ZQQG2  1384719342   
4   A94QU4C90B1AX  1384719342   

                                       reviewerName   helpful  overall  \
0  cassandra tu "Yeah, well, that's just like, u...    [0, 0]        5   
1                                              Jake  [13, 14]        5   
2                     Rick Bennette "Rick Bennette"    [1, 1]        5   
3                         RustyBill "Sunday Rocker"    [0, 0]        5   
4                                     SEAN MASLANKA    [0, 0]        5   

   unixReviewTime   reviewTime  \
0      1393545600  02 28, 2014   
1      1363392000  03 16, 2013   
2      1377648000  08 28, 2013   
3      1392336000  02 14, 2014   
4      1392940800  02 21, 2014   

                                             reviews sentiment  
0  good. Not much to write about here, but it doe...  positive  
1  Jake. The product does exactly as it should an...  positive  
2  It Does The Job Well. The primary job of this ...  positive  
3  GOOD WINDSCREEN FOR THE MONEY. Nice windscreen...  positive  
4  No more pops when I record my vocals.. This po...  positive

In [12]:
df['sentiment'].value_counts()

sentiment
positive    9022
negative    1239
Name: count, dtype: int64

In [23]:
import re, string
def f(text):
    text = str(text).lower()
    text = re.sub('\[.*?\]', '', text)
    text = re.sub('https?://\S+|www\.\S+', '', text)
    text = re.sub('<.*?>+', '', text)
    text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
    text = re.sub('\n', '', text)
    text = re.sub('\w*\d\w*', '', text)
    return text

df['reviews']=df['reviews'].apply(lambda x: f(x))
df.head()

reviewerID        asin  \
0  A2IBPI20UZIR0U  1384719342   
1  A14VAT5EAX3D9S  1384719342   
2  A195EZSQDW3E21  1384719342   
3  A2C00NNG1ZQQG2  1384719342   
4   A94QU4C90B1AX  1384719342   

                                       reviewerName   helpful  overall  \
0  cassandra tu "Yeah, well, that's just like, u...    [0, 0]        5   
1                                              Jake  [13, 14]        5   
2                     Rick Bennette "Rick Bennette"    [1, 1]        5   
3                         RustyBill "Sunday Rocker"    [0, 0]        5   
4                                     SEAN MASLANKA    [0, 0]        5   

   unixReviewTime   reviewTime  \
0      1393545600  02 28, 2014   
1      1363392000  03 16, 2013   
2      1377648000  08 28, 2013   
3      1392336000  02 14, 2014   
4      1392940800  02 21, 2014   

                                             reviews sentiment  
0  good not much to write about here but it does ...  positive  
1  jake the product does exactly as it should and...  positive  
2  it does the job well the primary job of this d...  positive  
3  good windscreen for the money nice windscreen ...  positive  
4  no more pops when i record my vocals this pop ...  positive

停用词的处理——去掉负面情绪的选择

In [24]:
stop_words= ['yourselves', 'between', 'whom', 'itself', 'is', "she's", 'up', 'herself', 'here', 'your', 'each', 
             'we', 'he', 'my', "you've", 'having', 'in', 'both', 'for', 'themselves', 'are', 'them', 'other',
             'and', 'an', 'during', 'their', 'can', 'yourself', 'she', 'until', 'so', 'these', 'ours', 'above', 
             'what', 'while', 'have', 're', 'more', 'only', "needn't", 'when', 'just', 'that', 'were', "don't", 
             'very', 'should', 'any', 'y', 'isn', 'who',  'a', 'they', 'to', 'too', "should've", 'has', 'before',
             'into', 'yours', "it's", 'do', 'against', 'on',  'now', 'her', 've', 'd', 'by', 'am', 'from', 
             'about', 'further', "that'll", "you'd", 'you', 'as', 'how', 'been', 'the', 'or', 'doing', 'such',
             'his', 'himself', 'ourselves',  'was', 'through', 'out', 'below', 'own', 'myself', 'theirs', 
             'me', 'why', 'once',  'him', 'than', 'be', 'most', "you'll", 'same', 'some', 'with', 'few', 'it',
             'at', 'after', 'its', 'which', 'there','our', 'this', 'hers', 'being', 'did', 'of', 'had', 'under',
             'over','again', 'where', 'those', 'then', "you're", 'i', 'because', 'does', 'all']
stop_words = set(stop_words)

In [25]:
df['reviews'] = df['reviews'].apply(lambda x: ' '.join([word for word in x.split() if word not in stop_words]))
df.head()

reviewerID        asin  \
0  A2IBPI20UZIR0U  1384719342   
1  A14VAT5EAX3D9S  1384719342   
2  A195EZSQDW3E21  1384719342   
3  A2C00NNG1ZQQG2  1384719342   
4   A94QU4C90B1AX  1384719342   

                                       reviewerName   helpful  overall  \
0  cassandra tu "Yeah, well, that's just like, u...    [0, 0]        5   
1                                              Jake  [13, 14]        5   
2                     Rick Bennette "Rick Bennette"    [1, 1]        5   
3                         RustyBill "Sunday Rocker"    [0, 0]        5   
4                                     SEAN MASLANKA    [0, 0]        5   

   unixReviewTime   reviewTime  \
0      1393545600  02 28, 2014   
1      1363392000  03 16, 2013   
2      1377648000  08 28, 2013   
3      1392336000  02 14, 2014   
4      1392940800  02 21, 2014   

                                             reviews sentiment  
0  good not much write but exactly supposed filte...  positive  
1  jake product exactly quite affordablei not rea...  positive  
2  job well primary job device block breath would...  positive  
3  good windscreen money nice windscreen protects...  positive  
4  no pops record vocals pop filter great looks p...  positive

In [ ]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()
df['sentiment'] = le.fit_transform(df['sentiment'])
df.head()

reviewerID        asin  \
0  A2IBPI20UZIR0U  1384719342   
1  A14VAT5EAX3D9S  1384719342   
2  A195EZSQDW3E21  1384719342   
3  A2C00NNG1ZQQG2  1384719342   
4   A94QU4C90B1AX  1384719342   

                                       reviewerName   helpful  overall  \
0  cassandra tu "Yeah, well, that's just like, u...    [0, 0]        5   
1                                              Jake  [13, 14]        5   
2                     Rick Bennette "Rick Bennette"    [1, 1]        5   
3                         RustyBill "Sunday Rocker"    [0, 0]        5   
4                                     SEAN MASLANKA    [0, 0]        5   

   unixReviewTime   reviewTime  \
0      1393545600  02 28, 2014   
1      1363392000  03 16, 2013   
2      1377648000  08 28, 2013   
3      1392336000  02 14, 2014   
4      1392940800  02 21, 2014   

                                             reviews  sentiment  
0  good not much write but exactly supposed filte...          1  
1  jake product exactly quite affordablei not rea...          1  
2  job well primary job device block breath would...          1  
3  good windscreen money nice windscreen protects...          1  
4  no pops record vocals pop filter great looks p...          1

In [28]:
df1 = df.copy()
df1 = df1[['reviews', 'sentiment', 'overall', 'reviewerID']].reset_index(drop=True)
df1.head()

reviews  sentiment  overall  \
0  good not much write but exactly supposed filte...          1        5   
1  jake product exactly quite affordablei not rea...          1        5   
2  job well primary job device block breath would...          1        5   
3  good windscreen money nice windscreen protects...          1        5   
4  no pops record vocals pop filter great looks p...          1        5   

       reviewerID  
0  A2IBPI20UZIR0U  
1  A14VAT5EAX3D9S  
2  A195EZSQDW3E21  
3  A2C00NNG1ZQQG2  
4   A94QU4C90B1AX

In [29]:
from nltk.stem.porter import PorterStemmer

ps = PorterStemmer()
corpus = []
for i in range(0, len(df1)):
    review = re.sub('[^a-zA-Z]', ' ', df1['reviews'][i])
    review = review.split()
    review = [ps.stem(word) for word in review if not word in stop_words]
    review = ' '.join(review)
    corpus.append(review)


In [30]:
corpus[0]

'good not much write but exactli suppos filter pop sound record much crisp one lowest price pop filter amazon might well buy honestli work despit price'

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(2, 2))
X = tfidf_vectorizer.fit_transform(df1['reviews'])
X.shape

(10261, 5000)

In [36]:
y = df['sentiment']

In [37]:
from collections import Counter
from imblearn.over_sampling import SMOTE

print(f'原始数据集形状: {Counter(y)}')
smote = SMOTE(random_state=42)
X_smote, y_smote = smote.fit_resample(X, y)
print(f'SMOTE后数据集形状: {Counter(y_smote)}')

原始数据集形状: Counter({1: 9022, 0: 1239})
SMOTE后数据集形状: Counter({1: 9022, 0: 9022})


In [39]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_smote, y_smote, test_size=0.2, random_state=42)

建立逻辑回归模型


In [40]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import cross_val_score

logreg_cv = LogisticRegression(random_state=0)
dt_cv = DecisionTreeClassifier()
knn_cv = KNeighborsClassifier()
svc_cv = SVC()
nb_cv = BernoulliNB()
cv_dict = {0: 'Logistic Regression', 1: 'Decision Tree', 2: 'KNN', 3: 'SVC', 4: 'Naive Bayes'}
cv_models = [logreg_cv, dt_cv, knn_cv, svc_cv, nb_cv]

for i,model in enumerate(cv_models):
    print("{} Test Accuracy: {}".format(cv_dict[i], cross_val_score(model, X, y, cv=10, scoring ='accuracy').mean()))

Logistic Regression Test Accuracy: 0.8840272676715049
Decision Tree Test Accuracy: 0.8356862756263155
KNN Test Accuracy: 0.8685310457795469
SVC Test Accuracy: 0.8850992026208548
Naive Bayes Test Accuracy: 0.8336402512285257


In [44]:
from sklearn.model_selection import GridSearchCV
import numpy as np

param_grid = {'C': np.logspace(-4, 4, 50),
             'penalty':['l2']}
clf = GridSearchCV(LogisticRegression(random_state=0), param_grid, cv=5, verbose=0, n_jobs=-1)
best_model = clf.fit(X_train,y_train)
print(best_model.best_estimator_)
print("The mean accuracy of the model is:",best_model.score(X_test,y_test))

LogisticRegression(C=10000.0, random_state=0)
The mean accuracy of the model is: 0.9332224993072873


In [45]:
logreg = LogisticRegression(C=10000.0, random_state=0)
logreg.fit(X_train, y_train)
y_pred = logreg.predict(X_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(X_test, y_test)))

Accuracy of logistic regression classifier on test set: 0.93


In [46]:
from sklearn.metrics import classification_report, confusion_matrix

y_pred = clf.predict(X_test)
print("Confusion Matrix:\n",confusion_matrix(y_test, y_pred))
print("Classification Report:\n",classification_report(y_test, y_pred))

Confusion Matrix:
 [[1791    5]
 [ 236 1577]]
Classification Report:
               precision    recall  f1-score   support

           0       0.88      1.00      0.94      1796
           1       1.00      0.87      0.93      1813

    accuracy                           0.93      3609
   macro avg       0.94      0.93      0.93      3609
weighted avg       0.94      0.93      0.93      3609



深度学习

In [3]:
import pandas as pd

df = pd.read_csv('Musical_instruments_reviews.csv')
df['reviewText'] = df['reviewText'].fillna('Missing')
df['reviews'] = df['summary'] + '. ' + df['reviewText']
df = df.drop(['summary','reviewText', 'reviewerName'], axis=1)
df.head()

reviewerID        asin   helpful  overall  unixReviewTime   reviewTime  \
0  A2IBPI20UZIR0U  1384719342    [0, 0]        5      1393545600  02 28, 2014   
1  A14VAT5EAX3D9S  1384719342  [13, 14]        5      1363392000  03 16, 2013   
2  A195EZSQDW3E21  1384719342    [1, 1]        5      1377648000  08 28, 2013   
3  A2C00NNG1ZQQG2  1384719342    [0, 0]        5      1392336000  02 14, 2014   
4   A94QU4C90B1AX  1384719342    [0, 0]        5      1392940800  02 21, 2014   

                                             reviews  
0  good. Not much to write about here, but it doe...  
1  Jake. The product does exactly as it should an...  
2  It Does The Job Well. The primary job of this ...  
3  GOOD WINDSCREEN FOR THE MONEY. Nice windscreen...  
4  No more pops when I record my vocals.. This po...

In [4]:
def func(row):
    if row['overall'] >= 4.0:
        return 'positive'
    else:
        return 'negative'

df['sentiment'] = df.apply(func, axis=1)
df['sentiment'] = df['sentiment'].map({'negative': 0, 'positive': 1})
df.head()

reviewerID        asin   helpful  overall  unixReviewTime   reviewTime  \
0  A2IBPI20UZIR0U  1384719342    [0, 0]        5      1393545600  02 28, 2014   
1  A14VAT5EAX3D9S  1384719342  [13, 14]        5      1363392000  03 16, 2013   
2  A195EZSQDW3E21  1384719342    [1, 1]        5      1377648000  08 28, 2013   
3  A2C00NNG1ZQQG2  1384719342    [0, 0]        5      1392336000  02 14, 2014   
4   A94QU4C90B1AX  1384719342    [0, 0]        5      1392940800  02 21, 2014   

                                             reviews  sentiment  
0  good. Not much to write about here, but it doe...          1  
1  Jake. The product does exactly as it should an...          1  
2  It Does The Job Well. The primary job of this ...          1  
3  GOOD WINDSCREEN FOR THE MONEY. Nice windscreen...          1  
4  No more pops when I record my vocals.. This po...          1

In [5]:
from sklearn.model_selection import train_test_split

train_texts, val_texts, train_labels, val_labels = train_test_split(
    df['reviews'].values,
    df['sentiment'].values,
    test_size=0.2,
    random_state=42
)

In [6]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader

class MyDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_len = max_len

    def __len__(self):
        return len(self.texts)
    
    def __getitem__(self, idx):
        text = str(self.texts[idx])
        label = self.labels[idx]
        encoding = self.tokenizer.encode_plus(
            text,
            add_special_tokens=True,
            max_length=self.max_len,
            truncation=True,
            padding='max_length',
            return_attention_mask=True,
            return_tensors='pt',
        )
        return {
            'text': text,
            'input_ids': encoding['input_ids'].flatten(),
            'attention_mask': encoding['attention_mask'].flatten(),
            'label': torch.tensor(label, dtype=torch.long)
        }

In [12]:
from transformers import BertTokenizer, BertModel

MAX_LEN = 128
BATCH_SIZE = 16
MODEL_NAME = "google-bert/bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(MODEL_NAME)
model = BertModel.from_pretrained(MODEL_NAME)

train_dataset = MyDataset(train_texts, train_labels, tokenizer, max_len=MAX_LEN)
train_data_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)

test_dataset = MyDataset(val_texts, val_labels, tokenizer, max_len=MAX_LEN)
test_data_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False)

batch_data = next(iter(train_data_loader))
print(batch_data)

d:\PythonVersions\python311\Lib\site-packages\transformers\tokenization_utils_base.py:1602: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be deprecated in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


{'text': ['Great for Drop Bb. Put my Schecter C-1 Custom FR in Drop Bb, these strings sound great and have a great low end. Highly reccomend.', 'nice and easy. Its nice and easy to use.  Secures the strings very good and works for all types of guitars with no issues.', "Great delay pedal. I really like this delay. It's simple and works well. I really am starting to love Behringer. Great pedal for a great price.", 'They slip out. Like many say, they do slip out easy so you need to keep that in mind, otherwise for the price and overall quality they are good cables.', 'No good. Arrive, in trashed packing, but the worst was that it sounded like a piece of chalk on a blackboard... certainly NOT professional quality or any quality that I would give it.', "Great tuner; can't beat the price!. Snark tuners are great! They are very accurate and are very inexpensive. I have now four of them so I can keep one in each guitar case and one for the uke!", "Great cheap pedal, but a little redundant for

In [14]:
class MyModel(nn.Module):
    def __init__(self, model, n_classes):
        super(MyModel, self).__init__()
        self.bert = model
        self.dropout = nn.Dropout(0.3)
        self.fc = nn.Linear(self.bert.config.hidden_size, n_classes)

    def forward(self, input_ids, attention_mask):
        output = self.bert(input_ids=input_ids, attention_mask=attention_mask)
        pooled_output = output.pooler_output
        output = self.dropout(pooled_output)
        output = self.fc(output)
        return output
    
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print("device: ", device)

mymodel = MyModel(model, 2)
mymodel.to(device)

device:  cuda


MyModel(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affin

In [20]:
from sklearn.metrics import classification_report
from transformers import get_linear_schedule_with_warmup
import numpy as np

def train_model(model, train_loader, optimizer, criterion, device, scheduler):
    model.train()
    losses = []
    corr_preds = 0

    for batch in train_loader:
        inputs_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['label'].to(device)

        optimizer.zero_grad()
        outputs = model(inputs_ids, attention_mask=attention_mask)

        _, preds = torch.max(outputs, dim=1)
        loss = criterion(outputs, labels)
        corr_preds += torch.sum(preds == labels)
        losses.append(loss.item())

        loss.backward()
        nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        optimizer.step()
        scheduler.step()
    return np.mean(losses), corr_preds.double() / len(train_loader.dataset)

def eval_model(model, val_loader, criterion, device):
    model.eval()
    losses = []
    corr_preds = 0
    all_preds = []
    all_labels = []

    with torch.no_grad():
        for batch in val_loader:
            inputs_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['label'].to(device)

            outputs = model(inputs_ids, attention_mask=attention_mask)
            _, preds = torch.max(outputs, dim=1)
            loss = criterion(outputs, labels)
            corr_preds += torch.sum(preds == labels)
            losses.append(loss.item())

            all_preds.extend(preds.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())
    
    acc = corr_preds.double() / len(val_loader.dataset)
    report = classification_report(all_labels, all_preds, target_names=['negative', 'positive'])
    return np.mean(losses), acc, report

EPOCHS = 5
optimizer = optim.AdamW(model.parameters(), lr=2e-5)
total_steps = len(train_data_loader) * EPOCHS
scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=0, num_training_steps=total_steps)
criterion = nn.CrossEntropyLoss()

import time
for epoch in range(EPOCHS):
    print(f'Epoch {epoch+1}/{EPOCHS}')
    print('-' * 10)

    start_time = time.time()
    train_loss, train_acc = train_model(mymodel, train_data_loader, optimizer, criterion, device, scheduler)
    val_loss, val_acc, val_report = eval_model(mymodel, test_data_loader, criterion, device)

    end_time = time.time()
    epoch_mins, epoch_secs = divmod(end_time - start_time, 60)
    print(f'Time: {epoch_mins}m {epoch_secs}s')
    print(f'Train Loss: {train_loss:.4f} | Train Acc: {train_acc*100:.4f}%')
    print(f'Val Loss: {val_loss:.4f} | Val Acc: {val_acc*100:.4f}%')

Epoch 1/5
----------
Time: 2.0m 11.715480089187622s
Train Loss: 0.0981 | Train Acc: 97.5634%
Val Loss: 0.2839 | Val Acc: 93.4730%
Epoch 2/5
----------
Time: 2.0m 12.320727348327637s
Train Loss: 0.0541 | Train Acc: 98.6355%
Val Loss: 0.3163 | Val Acc: 93.3268%
Epoch 3/5
----------
Time: 2.0m 12.040659666061401s
Train Loss: 0.0218 | Train Acc: 99.5249%
Val Loss: 0.4220 | Val Acc: 93.3268%
Epoch 4/5
----------
Time: 2.0m 11.8428316116333s
Train Loss: 0.0098 | Train Acc: 99.7807%
Val Loss: 0.4510 | Val Acc: 93.1807%
Epoch 5/5
----------
Time: 2.0m 11.332291841506958s
Train Loss: 0.0036 | Train Acc: 99.9147%
Val Loss: 0.4824 | Val Acc: 93.1807%


In [22]:
val_acc, val_loss, val_report = eval_model(
    mymodel,
    test_data_loader,
    criterion,
    device
)

print('Final Evaluation: ')
print(f'Loss: {val_acc:.4f}')
print(f'Accuracy: {val_loss:.4f}')
print('Classification Report:')
print(val_report)

Final Evaluation: 
Loss: 0.4824
Accuracy: 0.9318
Classification Report:
              precision    recall  f1-score   support

    negative       0.74      0.66      0.70       244
    positive       0.95      0.97      0.96      1809

    accuracy                           0.93      2053
   macro avg       0.85      0.81      0.83      2053
weighted avg       0.93      0.93      0.93      2053



- 传统机器学习方法经分词，去标点，tfidf 向量化，去除停用词及 SMOTE 过采样后，准确率在88%左右，经网格搜索调参后，准确率提升至93%左右
- 深度学习方法使用模型 `bert-base-uncased`，未进行其他处理，训练发现训练多个 epoch 有些许过拟合现象产生，可能由于negative样本较少，未经数据增强，总体准确率93%
- 二者总体效果接近，深度学习方法在未经数据增强的情况下，对negative样本的识别效果略差，传统机器学习方法在增加数据增强后，对两标签的识别效果均较好